src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -788,19 +781,60 @@ private Task WaitConnectionLockAsync([CallerMemberName] string memberName = null
         private class Subscription : IDisposable
         {
             private readonly InvocationHandler _handler;
-            private readonly List<InvocationHandler> _handlerList;
+            private readonly InvocationHandlerList _handlerList;
 
-            public Subscription(InvocationHandler handler, List<InvocationHandler> handlerList)
+            public Subscription(InvocationHandler handler, InvocationHandlerList handlerList)
             {
                 _handler = handler;
                 _handlerList = handlerList;
             }
 
             public void Dispose()
             {
-                lock (_handlerList)
+                 _handlerList.Remove(_handler);
+            }
+        }
+
+        private class InvocationHandlerList
+        {
+            private readonly List<InvocationHandler> _invocationHandlers;
+            private InvocationHandler[] _CopiedHandlers;
+
+            internal InvocationHandlerList(InvocationHandler handler)
+            {
+                _invocationHandlers = new List<InvocationHandler>() { handler };
+            }
+
+            internal InvocationHandler[] GetCopiedHandlers()
+            {
+                if (_CopiedHandlers == null)
+                {
+                    lock (_invocationHandlers)
+                    {
+                        _CopiedHandlers = _invocationHandlers.ToArray();
```

Comment:
1. Capture `_CopiedHandlers` into a local before testing against `null`, to avoid another thread messing with it.
2. Check `_CopiedHandlers` again when inside the lock to avoid another thread setting it after your null check.

So, something like this:

```csharp
// First, just grab the current value without the lock
var handlers = _CopiedHandlers;

// If handlers is non-null, we can just return it.
if(handlers == null) 
{
    lock(_invocationHandlers)
    {
        // Did another thread already set it?
        if(_CopiedHandlers == null)
        {
            // No, so initialize it
            _CopiedHandlers = _invocationHandlers.ToArray();
        }
        // Either way, copy the (now initialized) value locally while inside the lock
        handlers = _CopiedHandlers;
    }
}
return handlers;
```

That way, the fast path, when handlers are being retrieved and nobody is messing with them, will occur outside the lock, as desired, but when we do need to take the lock, we do everything inside the lock before releasing it.